src/featureforest/utils/data.py

```diff
@@ -1,106 +1,259 @@
     """
-    b, c, img_height, img_width = imgs.shape
-    margin = (patch_size - target_size) // 2
     new_width = img_width + (2 * margin)
     new_height = img_height + (2 * margin)
     pad_right = patch_size - (new_width % patch_size)
     pad_bottom = patch_size - (new_height % patch_size)
+
+    return pad_right, pad_bottom
+
+
+def patchify(
+    images: Tensor,
+    patch_size: Optional[int] = None,
+    overlap: Optional[int] = None
+) -> Tensor:
+    """Divide images into patches.
+    images: (B, C, H, W)
+    out: (B*N, C, patch_size, patch_size)
+
+    Args:
+        images (Tensor): a batch of images of shape (B, C, H, W)
+        patch_size (Optional[int], optional): patch size. Defaults to None.
+        overlap (Optional[int], optional): patch overlap. Defaults to None.
+
+    Returns:
+        Tensor: patches of the input batch of shape (B*N, C, patch_size, patch_size)
+    """
+    _, c, img_height, img_width = images.shape
+    if patch_size is None:
+        patch_size = get_patch_size(img_height, img_width)
+        overlap = 3 * patch_size // 4
+    if overlap is None:
+        overlap = 3 * patch_size // 4
+
+    stride, margin = get_stride_margin(patch_size, overlap)
+    pad_right, pad_bottom = get_paddings(patch_size, margin, img_height, img_width)
     pad = (margin, pad_right + margin, margin, pad_bottom + margin)
-    padded_imgs = F.pad(imgs, pad=pad, mode="reflect")
+    padded_imgs = F.pad(images, pad=pad, mode="reflect")
+    # making patches using torch unfold method
     patches = padded_imgs.unfold(
-        2, patch_size, step=target_size).unfold(
-        3, patch_size, step=target_size)
+        2, patch_size, step=stride).unfold(
+        3, patch_size, step=stride)
     patches = patches.permute(0, 2, 3, 1, 4, 5).reshape(
         -1, c, patch_size, patch_size
     )
 
     return patches
 
 
-def get_num_target_patches(img_height, img_width, patch_size, target_size):
-    margin = (patch_size - target_size) // 2
-    new_width = img_width + (2 * margin)
-    new_height = img_height + (2 * margin)
-    pad_right = patch_size - (new_width % patch_size)
-    pad_bottom = patch_size - (new_height % patch_size)
-    num_patches_w = int((img_width + pad_right) / target_size)
-    num_patches_h = int((img_height + pad_bottom) / target_size)
+def get_num_patches(
+    img_height: float, img_width: float, patch_size: int, overlap: int
+) -> Tuple[int, int]:
+    """Returns number of patches per each image dimension.
+
+    Args:
+        img_height (int): image height
+        img_width (int): image width
+        patch_size (int): patch size
+        overlap (int): patch overlap size
+
+    Returns:
+        Tuple[int, int]: number of patches for height and width of image
+    """
+    stride, margin = get_stride_margin(patch_size, overlap)
+    pad_right, pad_bottom = get_paddings(patch_size, margin, img_height, img_width)
+    num_patches_w = int((img_width + pad_right) / stride)
```

Comment:
Does this need rounding up if the value is not an integer?

Reply:
Basically, the number of patches should be integer if the right amount of pads are added.  
I've added `assert` check to make sure it is happening.